src/lib/ChatCompletionStream.ts

```diff
@@ -0,0 +1,311 @@
+    const pushQueue: ChatCompletionChunk[] = [];
+    const readQueue: ((chunk: ChatCompletionChunk | undefined) => void)[] = [];
+    let done = false;
+
+    this.on('chunk', (chunk) => {
+      const reader = readQueue.shift();
+      if (reader) {
+        reader(chunk);
+      } else {
+        pushQueue.push(chunk);
+      }
+    });
+
+    this.on('end', () => {
+      done = true;
+      for (const reader of readQueue) {
+        reader(undefined);
+      }
+      readQueue.length = 0;
+    });
+
+    return {
+      next: async (): Promise<IteratorResult<ChatCompletionChunk>> => {
+        if (!pushQueue.length) {
+          if (done) {
+            return { value: undefined, done: true };
+          }
+          return new Promise<ChatCompletionChunk | undefined>((resolve) => readQueue.push(resolve)).then(
+            (chunk) => (chunk ? { value: chunk, done: false } : { value: undefined, done: true }),
+          );
+        }
+        const chunk = pushQueue.shift()!;
+        return { value: chunk, done: false };
+      },
+    };
+  }
+
+  toReadableStream(): ReadableStream {
+    const stream = new Stream(this[Symbol.asyncIterator].bind(this), this.controller);
+    return stream.toReadableStream();
+  }
+}
+
+function finalizeChatCompletion(snapshot: ChatCompletionSnapshot): ChatCompletion {
+  const { id, choices, created, model } = snapshot;
+  return {
+    id,
+    choices: choices.map(({ message, finish_reason, index }): ChatCompletion.Choice => {
+      if (!finish_reason) throw new OpenAIError(`missing finish_reason for choice ${index}`);
+      const { content = null, function_call, role } = message;
+      if (!role) throw new OpenAIError(`missing role for choice ${index}`);
+      if (function_call) {
+        const { arguments: args, name } = function_call;
+        if (args == null) throw new OpenAIError(`missing function_call.arguments for choice ${index}`);
+        if (!name) throw new OpenAIError(`missing function_call.name for choice ${index}`);
+        return { message: { content, function_call: { arguments: args, name }, role }, finish_reason, index };
+      }
+      return { message: { content: content, role }, finish_reason, index };
+    }),
+    created,
+    model,
+    object: 'chat.completion',
+  };
+}
+
+/**
+ * Represents a streamed chunk of a chat completion response returned by model,
+ * based on the provided input.
+ */
+export interface ChatCompletionSnapshot {
+  /**
+   * A unique identifier for the chat completion.
+   */
+  id: string;
+
+  /**
+   * A list of chat completion choices. Can be more than one if `n` is greater
+   * than 1.
+   */
+  choices: Array<ChatCompletionSnapshot.Choice>;
+
+  /**
+   * The Unix timestamp (in seconds) of when the chat completion was created.
+   */
+  created: number;
+
+  /**
+   * The model to generate the completion.
+   */
+  model: string;
+}
+
+export namespace ChatCompletionSnapshot {
```

Comment:
Are these auto generated or manually kept in sync?

Reply:
These are manually defined, everything in the `src/lib` folder is not touched by codegen and is manually written on our end.